### PR TITLE
Select concrete class for abstract types

### PR DESCRIPTION
**Pull request by thk123**
Issue: diffblue/test-gen#402

This builds on the the work of #1084 and uses the information of abstract classes to select concrete subtype. :white_check_mark: 

The Java tests can be seen in the corresponding test-gen PR, but I have also added tests to CBMC to demonstrate the modifications to the generated nondet initialisation code. For example, if you have an interface `I` with implementation `A`, the nondet code would look something like:

```C
struct A { int a } object_factory$1
A* object_factory$2
if(NONDET(bool))
{
  object_factory$2 = NULL
}
else
{
  object_factory$2 = &object_factory$1;
  object_factory$1.a = NONDET(int);
}
I *param_0 = (I *)object_factory2;
```

test-gen PR to go along with this: diffblue/test-gen#706

### CODE REVIEW

File: regression/cbmc-java/instance-method-user-interface-field/TestClass.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    public static void main(String[] args)
+    {
+        // Ensure A, B, C, D are converted into GOTO
+        A a = new A();
+        B b = new B();
+
+        a.getANumber();
```
Comment (thk123) [inline regression/cbmc-java/instance-method-user-interface-field/TestClass.java]: These calls aren't required

File: regression/cbmc-java/instance-method-user-interface-field/test.desc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CORE
+TestClass.class
+--cover location --function TestClass.foo --show-goto-functions
+
+\.someObject = \(struct I \*\)\w+
+@class_identifier = "java::A";
+--
```
Comment (thk123) [inline regression/cbmc-java/instance-method-user-interface-field/test.desc]: Missing new line

File: regression/cbmc-java/instance-method-user-interface-two-fields-different/TestClass.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    public static void main(String[] args)
+    {
+        // Ensure A, B, C, D are converted into GOTO
+        A a = new A();
+        B b = new B();
+        C c = new C();
+        D d = new D();
+
+        a.getANumber();
```
Comment (thk123) [inline regression/cbmc-java/instance-method-user-interface-two-fields-different/TestClass.java]: These calls aren't required

File: regression/cbmc-java/static-method-user-interface-param/TestClass.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        }
+    }
+
+    public static void main(String[] args)
+    {
+        // Ensure A, B, C, D are converted into GOTO
+        A a = new A();
+        B b = new B();
+
+        a.getANumber();
```
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-param/TestClass.java]: These calls aren't required

File: regression/cbmc-java/static-method-user-interface-param/test.desc [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CORE
+TestClass.class
+--cover location --function TestClass.foo --show-goto-functions
+
+@class_identifier = "java::A";
+ = \(struct I \*\)
+tmp_object_factory\$\d+ = null;
```
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-param/test.desc]: Missing new line

File: regression/cbmc-java/static-method-user-interface-two-params-different/TestClass.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    public static void main(String[] args)
+    {
+        // Ensure A, B, C, D are converted into GOTO
+        A a = new A();
+        B b = new B();
+        C c = new C();
+        D d = new D();
+
+        a.getANumber();
```
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-two-params-different/TestClass.java]: These calls are not needed

File: regression/cbmc-java/static-method-user-interface-two-params-different/test.desc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+CORE
+TestClass.class
+--cover location --function TestClass.foo --show-goto-functions
+
+@class_identifier = "java::A";
+ = \(struct I \*\)
+@class_identifier = "java::C";
+ = \(struct J \*\)
+tmp_object_factory\$\d+ = null;
```
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-two-params-different/test.desc]: Missing new line

File: regression/cbmc-java/static-method-user-interface-two-params-same/TestClass.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    public static void main(String[] args)
+    {
+        // Ensure A, B, C, D are converted into GOTO
+        A a = new A();
+        B b = new B();
+
+        a.getANumber();
```
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-two-params-same/TestClass.java]: These calls are not needed

File: src/java_bytecode/java_object_factory.cpp [hunk 9]
```diff
@@ -621,6 +683,11 @@ void java_object_factoryt::gen_nondet_init(
 
   if(type.id()==ID_pointer)
   {
+
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Remove this unusued bracket

File: src/java_bytecode/java_object_factory.cpp [hunk 10]
```diff
@@ -494,6 +519,42 @@ void java_object_factoryt::gen_nondet_pointer_init(
...
+    "",
+    false,
+    create_dynamic_objects,
+    false,
+    typet(),
+    update_in_placet::NO_UPDATE_IN_PLACE);
+
+  return new_symbol.symbol_expr();
+}
+
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Too many spaces

File: src/java_bytecode/java_object_factory.cpp [hunk 11]
```diff
@@ -59,22 +61,16 @@ class java_object_factoryt
...
-  void gen_pointer_target_init(
-    const exprt &expr,
-    const typet &target_type,
-    bool create_dynamic_objects,
-    update_in_placet update_in_place);
 
   code_assignt get_null_assignment(
     const exprt &expr,
     const pointer_typet &ptr_type);
-
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: This shouldn't be removed

File: src/java_bytecode/java_object_factory.cpp [hunk 12]
```diff
@@ -831,6 +898,7 @@ void java_object_factoryt::gen_nondet_array_init(
   assignments.move_to_operands(init_done_label);
 }
 
+
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Unnecessary whitespace

File: src/java_bytecode/select_pointer_type.cpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+/// Find out whether a pointer is pointing to an abstract class or interface
+/// \param pointer_type: The type of the pointer
+/// \return True if the type pointed to is either an abstract class or an
+///   interface.
+bool select_pointer_typet::is_abstract_type(
+  const pointer_typet pointer_type) const
+{
+  const typet &pointing_to_type=ns.follow(pointer_type.subtype());
+
+  // void* pointers should be not considered abstract
```
Comment (thk123) [inline src/java_bytecode/select_pointer_type.cpp]: Explain why `void*` pointers can happen

File: regression/cbmc-java/static-method-user-interface-param/test.desc (hunk 5 above)
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-param/test.desc]: Fixed

File: regression/cbmc-java/static-method-user-interface-two-params-different/test.desc (hunk 7 above)
Comment (thk123) [inline regression/cbmc-java/static-method-user-interface-two-params-different/test.desc]: Fixed

File: src/java_bytecode/java_object_factory.cpp (hunk 10 above)
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Fixed

File: src/java_bytecode/select_pointer_type.cpp (hunk 13 above)
Comment (thk123) [inline src/java_bytecode/select_pointer_type.cpp]: Fixed

**Comment by thk123** [comment]
Some string tests are failing since there must be a pointer type whose subtype is neither empty nor a class. I suppose this is to do with how strings are represented. I suppose I can swap the `INVARIANT` to just a false since all other types are definitely _not_ abstract classes!

File: src/java_bytecode/get_concrete_class_alphabetically.cpp [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    std::back_inserter(concrete_childen),
+    [&] (const irep_idt &child_class_id)
+    {
+      const symbolt &type_symbol=symbol_table.lookup(child_class_id);
+      const class_typet &child_type=to_class_type(type_symbol.type);
+      return child_type.is_class() && !child_type.is_abstract();
+    });
+
+
+  if(concrete_childen.size()>0)
```
Comment (mgudemann) [inline src/java_bytecode/get_concrete_class_alphabetically.cpp]: please use `!..empty()`

File: src/java_bytecode/get_concrete_class_alphabetically.h [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+class get_concrete_class_alphabeticallyt
+{
+public:
+  explicit get_concrete_class_alphabeticallyt(const namespacet &ns):ns(ns)
+  {}
+
+  pointer_typet operator()(const pointer_typet &pointer_type);
+
+private:
+  const namespacet& ns;
```
Comment (mgudemann) [inline src/java_bytecode/get_concrete_class_alphabetically.h]: should be ` &ns`

File: src/java_bytecode/java_object_factory.cpp [hunk 16]
```diff
@@ -74,7 +71,7 @@ class java_object_factoryt
     const exprt &expr,
     const typet &target_type,
     bool create_dynamic_objects,
-    update_in_placet);
+    update_in_placet update_in_place);
```
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: why change this?

File: src/java_bytecode/select_pointer_type.cpp [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+  return resolved_type;
+}
+
+/// Find out whether a pointer is pointing to an abstract type (either an
+/// abstract class or an interface).
+/// \param pointer_type: The type of the pointer
+/// \return True if the type pointed to is either an abstract class or an
+///   interface.
+bool select_pointer_typet::is_abstract_type(
+  const pointer_typet pointer_type) const
```
Comment (mgudemann) [inline src/java_bytecode/select_pointer_type.cpp]: can be passed as reference

File: regression/cbmc-java/instance-method-user-interface-field/TestClass.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+                return false;
+            }
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    public static void main(String[] args)
```
Comment (mgudemann) [inline regression/cbmc-java/instance-method-user-interface-field/TestClass.java]: main function is not needed (in other tests, too)

Comment (mgudemann) [inline regression/cbmc-java/instance-method-user-interface-field/TestClass.java]: ok, understood, could you change the comment into something along the line
```
// ensure that A, B are referenced explicitly in order to get them converted into GOTO
```

File: src/java_bytecode/select_pointer_type.cpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+/// Select what type should be used for a given pointer type. If the class
+/// pointed to is abstract then we consider using a subclass of it.
+/// \param pointer_type: The pointer type replace
+/// \return A pointer type where the subtype may have been modified
+pointer_typet select_pointer_typet::operator()(
+  const pointer_typet &pointer_type) const
+{
+  // Currently only replace abstract classes
+  // Potentially in the future we might want to consider anything with
+  // derivied classes
```
Comment (mgudemann) [inline src/java_bytecode/select_pointer_type.cpp]: derived

File: src/java_bytecode/java_object_factory.cpp (hunk 16 above)
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Only reason would be greater consistency - can undo if you think it is a backward step?

**Comment by thk123** [comment]
@mgudemann Have done your requested changes - let me know if you want me to revert adding the variable name.

File: src/java_bytecode/java_object_factory.cpp (hunk 16 above)
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: no, leave it at it is
